confdgnmi/README.md

```diff
@@ -0,0 +1,283 @@
+
+## Run
+
+### Server
+                              
+Run server in demo mode (running ConfD not needed):   
+`./src/confd_gnmi_server.py` or `./src/confd_gnmi_server.py -t demo`
+
+Run server in ConfD api mode (running ConfD needed `make clean all start`):
+`./src/confd_gnmi_server.py -t api`
+
+Supported server options:
+ 
+```
+Usage: confd_gnmi_server.py [options]
+
+Options:
+  -h, --help            show this help message and exit
+  -t TYPE, --type=TYPE  gNMI server type  [api, netconf, demo]
+  --logging=LOGGING     Logging level [error, warning, info, debug]
+  -d CONFD_DEBUG, --confd-debug=CONFD_DEBUG
+                        ConfD debug level  [trace, debug, silent, proto]
+  --confd-addr=CONFD_ADDR
+                        ConfD IP address (default is 127.0.0.1)
+  --confd-port=CONFD_PORT
+                        ConfD port (default is 4565)
+```
+
+NOTE: `netconf` type server not yet implemented
+NOTE: Other parameters (e.g ports)  are currently hardcoded in source code (TODO)
+
+### Client
+
+(in other terminal than server)
+
+`./src/confd_gnmi_client.py`
+
+Supported options:
+
+```
+Usage: confd_gnmi_client.py [options]
+
+Options:
+  -h, --help            show this help message and exit
+  -o OPERATION, --oper=OPERATION
+                        gNMI operation [capabilities, set, get, subscribe]
+  --logging=LOGGING     Logging level [error, warning, info, debug]
+  --prefix=PREFIX       'prefix' path for set, get and subscribe operation
+                        (empty by default)
+  -p PATHS, --path=PATHS
+                        'path' for get, set and subscribe operation, can be
+                        repeated (empty by default)
+  -v VALS, --val=VALS   'value' for set operation, can be repeated (empty by
+                        default)
+```
+      
+Examples:
+
+`./src/confd_gnmi_client.py -o capabilities`
+`./src/confd_gnmi_client.py -o  get --prefix /interfaces --path interface[name=if_82]/name --path interface[name=if_82]/type`
+`./src/confd_gnmi_client.py -o set  --prefix /interfaces --path interface[name=if_82]/type --val fastEther`
+`./src/confd_gnmi_client.py -o subscribe --prefix /interfaces --path interface[name=if_82]/name --path interface[name=if_82]/type`
+
+NOTE: Other parameters (username, password, subscription types are currently hardcoded, TODO)
+    
+## Limitations and TODOs of current implementation
+
+* `encoding` - used only `BYTES`
+* `Get`, `Set` and `Subscribe` works only on `leaf` elements
+* `Subscribe` - only `ONCE` and `POLL` is implemented so far
+    * `updates_only` not supported
+    * `heartbeat_interval` not supported
+    * `sync_response` not generated
+* all values `TypedValues` are used as strings (`string_val`)
+* gNMI Path is converted to XPath or formatted path with simple string operations 
+  no datamodel knowledge used) 
+  
+## Investigations/TODOs
+
+TODO converting XPath to ConfD formatted path/keypath and back.
```

Comment:
I would primarily use RESTCONF instead of MAAPI or NETCONF

Reply:
Yes, RESTCONF adapter would be another option (or even MIXED adapter combining approaches). RESTCONF requires ConfD Premium.

Reply:
RESTCONF can be added in next milestones